### PR TITLE
Kandinsky 5 10 sec (NABLA suport)

### PR DESCRIPTION
**Pull request by leffff**
This PR adds support for 10 sec Kandinsky 5.0 model herd.

```python
import torch
from diffusers import Kandinsky5T2VPipeline
from diffusers.utils import export_to_video

# Load the pipeline
pipe = Kandinsky5T2VPipeline.from_pretrained(
    "ai-forever/Kandinsky-5.0-T2V-Lite-sft-10s-Diffusers", 
    torch_dtype=torch.bfloat16
)
pipe = pipe.to("cuda")

# Generate video
prompt = [
    "Photorealistic closeup video of two intricately detailed pirate ships locked in a fierce battle, complete with cannon fire and billowing sails, as they sail through the swirling waters of a steaming cup of coffee. The ships are miniature but highly realistic, with wooden textures and flags fluttering in the liquid breeze. Coffee splashes and foam ripple around them as they maneuver through the turbulent surface, dodging each other's attacks. A detailed reflection of the battle appears on the glossy surface of the coffee, adding to the dynamic realism. The camera pans and zooms to capture every dramatic moment of the high-seas clash within this tiny, unexpected world.",
    "Bad quality",
]
negative_prompt = "Static, 2D cartoon, cartoon, 2d animation, paintings, images, worst quality, low quality, ugly, deformed, walking backwards"

pipe.transformer.set_attention_backend("flex")

output = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    height=512,
    width=768,
    num_frames=241,
    num_inference_steps=50,
    guidance_scale=5.0,
    num_videos_per_prompt=1,
    generator=torch.Generator(42)
)
```

https://github.com/user-attachments/assets/52647681-0178-4797-88f3-de0506db5a3d

### CODE REVIEW

**Comment by leffff** [comment]
@yiyixuxu 

Plz check out this PR 


> Just commenting to note that we support [all kinds of different attention backends](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention_dispatch.py) now. So, as long as we implement the attention class in [this way](https://github.com/huggingface/diffusers/blob/4715c5c769e5ed2f3f174ee7aa76c67eb875b669/src/diffusers/models/transformers/transformer_flux.py#L275), for example, swapping a backend from SDPA ("native" in our terminology) to "flex" should be very easy.
> 
> ```python
> model.set_attention_backend("flex")
> ```

Yes, you are right. I tried doing 
```python
pipe.transformer.set_attention_backend("flex")
```
and it almost worked. You see, when I made separate processors, I did this: 
```python
class Kandinsky5NablaAttentionProcessor(nn.Module):
    """Custom attention processor for Nabla attention"""
    
    @torch.compile(mode="max-autotune-no-cudagraphs", dynamic=True)
    def __call__(
        self,
        attn,
        query,
        key,
        value,
        sparse_params=None,
        **kwargs,
    ):
        if sparse_params is None:
            raise ValueError("sparse_params is required for Nabla attention")

        query = query.transpose(1, 2).contiguous()
        key = key.transpose(1, 2).contiguous()
        value = value.transpose(1, 2).contiguous()

        block_mask = nablaT_v2(
            query,
            key,
            sparse_params["sta_mask"],
            thr=sparse_params["P"],
        )
        out = (
            flex_attention(query, key, value, block_mask=block_mask)
            .transpose(1, 2)
            .contiguous()
        )
        out = out.flatten(-2, -1)
        return out
```

**Comment by leffff** [comment]
Please check https://github.com/huggingface/diffusers/pull/12520/commits/b35445c65ab61f3d0e63b18967ca730757b28ca5

**Comment by MeiYi-dev** [comment]
> Please check [b35445c](https://github.com/huggingface/diffusers/commit/b35445c65ab61f3d0e63b18967ca730757b28ca5)

1: Unrelated to this PR, but what size will the PRO version of Kandinsky 5.0 be? I really like the 2B version, for it's size it's so good. I have a feeling if the PRO version is ~16B range it could possibly dethrone Waver 1.0 / Seedance 1.0 / Veo3 Fast on video arena (https://artificialanalysis.ai/text-to-video/arena?tab=leaderboard).

2: When can I expect the technical report to be available? Did you guys use REPA during training?

**Comment by sayakpaul** [comment]
I see what you mean @leffff! I don't think we should be compiling by default. Users can always take care of the compilation. 

So, essentially:

```py
pipe.transformer.set_attention_backend("flex")
pipe.transformer.compile(mode="max-autotune-no-cudagraphs", dynamic=True)
```

Notes:

* We should have `_repeated_blocks` configured for Kandinsky 5 to allow regional compilation. Refer https://huggingface.co/docs/diffusers/main/en/optimization/torch2.0#regional-compilation.
* Curious: why do we want to not allow CUDAgraphs here. I would assume "max-autotune" (which allows CUDAgraphing) should optimize the kernel launch overheads, no?

**Comment by leffff** [comment]
if users do not:
```
pipe.transformer.set_attention_backend("flex")
pipe.transformer.compile(mode="max-autotune-no-cudagraphs", dynamic=True)
```

10 sec model do not work. you are gonna get OOM 100%. so its mandatory

**Comment by yiyixuxu** [comment]
hi @leffff 
I agree with @sayak here. We shouldn't compile attention implicitly by default - I can see a lot of issues this could potentially cause, e.g. if users apply compile() at the transformer level, or try to use other stuff like regional compilation.

Instead, let's have users compile transformer themselves (regional compile might be most suitable here), which solves the OOM issue while giving them full control over compile settings. We should document this clearly and if you prefer, we add a warning if flex attention runs without any compilation.

check out regional compile here: https://huggingface.co/docs/diffusers/main/en/optimization/fp16#regional-compilation

**Comment by sayakpaul** [comment]
Strongly agree with @yiyixuxu in https://github.com/huggingface/diffusers/pull/12520#issuecomment-3428502112.

**Comment by leffff** [comment]
Okay, so @sayakpaul and @yiyixuxu propose, just adding this info to README of each model like so: 

```python
import torch
from diffusers import Kandinsky5T2VPipeline
from diffusers.utils import export_to_video

# Load the pipeline
pipe = Kandinsky5T2VPipeline.from_pretrained(
    "ai-forever/Kandinsky-5.0-T2V-Lite-sft-10s-Diffusers", 
    torch_dtype=torch.bfloat16
)
pipe = pipe.to("cuda")

# Generate video
prompt = [
    "Photorealistic closeup video of two intricately detailed pirate ships locked in a fierce battle, complete with cannon fire and billowing sails, as they sail through the swirling waters of a steaming cup of coffee. The ships are miniature but highly realistic, with wooden textures and flags fluttering in the liquid breeze. Coffee splashes and foam ripple around them as they maneuver through the turbulent surface, dodging each other's attacks. A detailed reflection of the battle appears on the glossy surface of the coffee, adding to the dynamic realism. The camera pans and zooms to capture every dramatic moment of the high-seas clash within this tiny, unexpected world.",
    "Bad quality",
]
negative_prompt = "Static, 2D cartoon, cartoon, 2d animation, paintings, images, worst quality, low quality, ugly, deformed, walking backwards"

pipe.transformer.set_attention_backend("flex")
pipe.transformer.compile_repeated_blocks(mode="max-autotune-no-cudagraphs", dynamic=True)

output = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    height=512,
    width=768,
    num_frames=241,
    num_inference_steps=50,
    guidance_scale=5.0,
    num_videos_per_prompt=1,
    generator=torch.Generator(42)
)
```

**Comment by sayakpaul** [comment]
Yes, that should cut it!

**Comment by leffff** [comment]
Okay, this seems to be working:
```python
import torch
from diffusers import Kandinsky5T2VPipeline
from diffusers.utils import export_to_video

# Load the pipeline
pipe = Kandinsky5T2VPipeline.from_pretrained(
    "ai-forever/Kandinsky-5.0-T2V-Lite-sft-10s-Diffusers", 
    torch_dtype=torch.bfloat16
)
pipe = pipe.to("cuda")

pipe.transformer.set_attention_backend("flex")
pipe.transformer.compile(mode="max-autotune-no-cudagraphs", dynamic=True)

# Generate video
prompt = "A cat and a dog baking a cake together in a kitchen."
negative_prompt = "Static, 2D cartoon, cartoon, 2d animation, paintings, images, worst quality, low quality, ugly, deformed, walking backwards"

output = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    height=512,
    width=768,
    num_frames=241,
    num_inference_steps=50,
    guidance_scale=5.0,
).frames[0]

## Save the video
export_to_video(output, "output.mp4", fps=24, quality=9)
```

from: https://huggingface.co/ai-forever/Kandinsky-5.0-T2V-Lite-sft-10s-Diffusers

**Comment by leffff** [comment]
@yiyixuxu @sayakpaul
Please review, but I believe, we good. If is ok for this PR - we can close it. 
Please check out this PR: https://github.com/huggingface/diffusers/pull/12527 we need to merge it ASAP.
Things left from me: Docs and Tests.

**Comment by sayakpaul** [comment]
@leffff let's add the tests and docs as well.

**Comment by yiyixuxu** [comment]
ok, let's just use this PR to add docs and tests?

**Comment by leffff** [comment]
Okay

**Comment by leffff** [comment]
Please checkout the docs

File: docs/source/en/api/pipelines/kandinsky_v5.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+> [!TIP]
+> Check out the [AI Forever](https://huggingface.co/ai-forever) organization on the Hub for the official model checkpoints for text-to-video generation, including pretrained, SFT, no-CFG, and distilled variants.
+
+## Available Models
+
+Kandinsky 5.0 T2V Lite comes in several variants optimized for different use cases:
+
+| Model Type | Description | Use Cases |
+|------------|-------------|-----------|
+| **SFT** | Supervised Fine-Tuned model | Highest generation quality |
```
Comment (yiyixuxu) [inline docs/source/en/api/pipelines/kandinsky_v5.md]: should we add model ids there?

File: docs/source/en/api/pipelines/kandinsky_v5.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    num_frames=121,  # ~5 seconds at 24fps
+    num_inference_steps=50,
+    guidance_scale=5.0,
+).frames[0]
+
+export_to_video(output, "output.mp4", fps=24, quality=9)
+```
+
+
+### Using Different Model Variants
```
Comment (yiyixuxu) [inline docs/source/en/api/pipelines/kandinsky_v5.md]: I think we should include an example for 10s and hightlight the fact that it needs to be compiled no?

**Comment by yiyixuxu** [comment]
@bot /style

File: docs/source/en/api/pipelines/kandinsky_v5.md (hunk 2 above)
Comment (leffff) [reply docs/source/en/api/pipelines/kandinsky_v5.md]: https://github.com/huggingface/diffusers/pull/12520/commits/dd6bf3982aa8991a2c74c4d44250e341a9b20c55

File: docs/source/en/api/pipelines/kandinsky_v5.md (hunk 1 above)
Comment (leffff) [reply docs/source/en/api/pipelines/kandinsky_v5.md]: https://github.com/huggingface/diffusers/pull/12520/commits/dd6bf3982aa8991a2c74c4d44250e341a9b20c55

**Comment by leffff** [comment]
@yiyixuxu plz check the new docs version!

**Review by yiyixuxu — APPROVED** [review]
looks really good! thanks!

**Comment by sayakpaul** [comment]
@leffff could you also add `kandinsky_v5` to `_toctree.yml`?

**Comment by leffff** [comment]
Okay!

**Comment by leffff** [comment]
@sayakpaul @yiyixuxu done!

**Comment by leffff** [comment]
Please review and merge!

File: src/diffusers/models/transformers/transformer_kandinsky.py [hunk 3]
```diff
@@ -324,6 +324,7 @@ def apply_rotary(x, rope):
                 sparse_params["sta_mask"],
                 thr=sparse_params["P"],
             )
+
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_kandinsky.py:327]: (nit): I guess we can discard all the changes in this file?

**Comment by leffff** [comment]
Okay!

File: src/diffusers/models/transformers/transformer_kandinsky.py (hunk 3 above)
Comment (leffff) [reply src/diffusers/models/transformers/transformer_kandinsky.py:327]: how do we do that?

**Comment by leffff** [comment]
Please check tests

File: tests/pipelines/kandinsky5/test_kandinsky5.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        
+        # Compare with output from prompt strings
+        inputs_with_prompt = self.get_dummy_inputs(torch_device)
+        output_with_prompt = pipe(**inputs_with_prompt).frames
+        
+        # Should be similar but not exactly the same due to different encoding
+        self.assertEqual(output_with_embeds.shape, output_with_prompt.shape)
+
+
+@slow
```
Comment (sayakpaul) [inline tests/pipelines/kandinsky5/test_kandinsky5.py]: We can leave this one out for now.

File: tests/pipelines/kandinsky5/test_kandinsky5.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+            pipe_loaded.to(torch_device)
+            pipe_loaded.set_progress_bar_config(disable=None)
+
+        inputs = self.get_dummy_inputs(torch_device)
+        output_loaded = pipe_loaded(**inputs).frames
+
+        max_diff = np.abs(output.detach().cpu().numpy() - output_loaded.detach().cpu().numpy()).max()
+        self.assertLess(max_diff, 1e-4)
+
+    def test_prompt_embeds(self):
```
Comment (sayakpaul) [inline tests/pipelines/kandinsky5/test_kandinsky5.py]: We should be able to test it with:
https://github.com/huggingface/diffusers/blob/dc6bd1511a4948ebca35b22609002bba58e71c83/tests/pipelines/test_pipelines_common.py#L2072

Comment (leffff) [reply tests/pipelines/kandinsky5/test_kandinsky5.py]: https://github.com/huggingface/diffusers/pull/12520/commits/56c4238615050fe7c011b2c23afc1372816da302

**Comment by leffff** [comment]
Can we merge?

**Comment by yiyixuxu** [comment]
@bot /style

**Comment by leffff** [comment]
Lets merge. We really want the docs to appear on the website.

**Comment by leffff** [comment]
When can we expect for the docs to appear here?
https://huggingface.co/docs/diffusers/index

**Comment by sayakpaul** [comment]
https://huggingface.co/docs/diffusers/main/en/api/pipelines/kandinsky5

**Comment by leffff** [comment]
Thank you!!!
